src/Npgsql/NpgsqlRawCopyStream.cs

```diff
@@ -152,25 +212,64 @@ Task FlushAsync(bool async)
 
         #region Read
 
-        public override int Read(byte[] buffer, int offset, int count) => Read(buffer, offset, count, false).GetAwaiter().GetResult();
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            ValidateArguments(buffer, offset, count);
+            return Read(new Span<byte>(buffer, offset, count));
+        }
 
         public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
         {
+            ValidateArguments(buffer, offset, count);
             if (cancellationToken.IsCancellationRequested)
                 return Task.FromCanceled<int>(cancellationToken);
             using (NoSynchronizationContextScope.Enter())
-                return Read(buffer, offset, count, true);
+                return ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
         }
 
-        async Task<int> Read(byte[] buffer, int offset, int count, bool async)
+#if !NET461 && !NETSTANDARD2_0
+        public override int Read(Span<byte> span)
+#else
+        public int Read(Span<byte> span)
+#endif
         {
             CheckDisposed();
             if (!CanRead)
                 throw new InvalidOperationException("Stream not open for reading");
 
-            if (_isConsumed) {
-                return 0;
+            var count = ReadCore(span.Length, false).GetAwaiter().GetResult();
+            if (count > 0)
+                _readBuf.ReadBytes(span.Slice(0, count));
+            return count;
+        }
+
+#if !NET461 && !NETSTANDARD2_0
+        public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+#else
+        public ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+#endif
+        {
+            CheckDisposed();
+            if (!CanRead)
+                throw new InvalidOperationException("Stream not open for reading");
+            if (cancellationToken.IsCancellationRequested)
+                return new ValueTask<int>(Task.FromCanceled<int>(cancellationToken));
+            using (NoSynchronizationContextScope.Enter())
+                return ReadLong(buffer);
+
+            async ValueTask<int> ReadLong(Memory<byte> buffer)
```

Comment:
ReadAsync always calls into ReadLong, so there no reason to keep the latter as a local function... Just make ReadAsync itself async and move the code there.

Reply:
Changed.